File: CS/PerformanceCounterCorrelationExample/Form1.cs
using System;
using System.Diagnostics;
using System.Windows.Forms;

namespace PerformanceCounterCorrelation {
    public partial class Form1 : Form {
        bool isRunning;
        PerformanceDataProvider dataProvider;
        WorksheetProcessor sheetProcessor;
        PerformanceCounter firstCounter = new PerformanceCounter();
        PerformanceCounter secondCounter = new PerformanceCounter();

        public Form1() {
            InitializeComponent();
            FillCounterInformation();
            this.btnStart.Select();
        }

        private void FillCounterInformation() {
            string processName = Process.GetCurrentProcess().ProcessName;
            this.textEditCategory1.Text = "Process";
            this.textEditCounter1.Text = "Private Bytes";
            this.textEditInstance1.Text = processName;
            this.textEditCategory2.Text = ".NET CLR Memory";
            this.textEditCounter2.Text = "# Bytes In All Heaps";
            this.textEditInstance2.Text = processName;
        }

        bool CreateCounters() {
            Process p = Process.GetCurrentProcess();
            try {
                firstCounter = new PerformanceCounter(textEditCategory1.Text, textEditCounter1.Text, textEditInstance1.Text);
                secondCounter = new PerformanceCounter(textEditCategory2.Text, textEditCounter2.Text, textEditInstance2.Text);
                firstCounter.NextValue();
                secondCounter.NextValue();
            }
            catch (Exception e) {
                MessageBox.Show(e.Message,"Counter Error");
                return false;
            }
            return true;
    }

        private void btnStart_Click(object sender, EventArgs e) {
            if (!isRunning) {
                if (!CreateCounters()) return;
                btnStart.Text = "Stop";
                dataProvider = new PerformanceDataProvider(firstCounter, secondCounter);
                dataProvider.DataReady += DataProvider_DataReady;
                dataProvider.DataOver += DataProvider_DataOver;
                isRunning = true;
                dataProvider.Start(500);
            }
            else {
                FullStop();
            }
        }

        private void DataProvider_DataReady(object sender, EventArgs e) {
            sheetProcessor = new WorksheetProcessor(dataProvider.GetData());
            arcScaleComponent1.DataBindings.Add("Value", sheetProcessor.GetDataSource(), "Column 0");
            labelComponent1.DataBindings.Add("Text", sheetProcessor.GetDataSource(), "Column 0");
        }

        private void DataProvider_DataOver(object sender, EventArgs e) {
            FullStop();
        }

        void FullStop() {
            dataProvider.DataReady -= DataProvider_DataReady;
            dataProvider.DataOver -= DataProvider_DataOver;
            btnStart.Text = "Start";
            dataProvider.Stop();
            sheetProcessor.SaveData("mytest");
            isRunning = false;
            arcScaleComponent1.DataBindings.Clear();
            labelComponent1.DataBindings.Clear();
        }


    }
}

File: VB/PerformanceCounterCorrelationExample/Form1.vb
Imports System
Imports System.Diagnostics
Imports System.Windows.Forms

Namespace PerformanceCounterCorrelation

    Public Partial Class Form1
        Inherits Form

        Private isRunning As Boolean

        Private dataProvider As PerformanceDataProvider

        Private sheetProcessor As WorksheetProcessor

        Private firstCounter As PerformanceCounter = New PerformanceCounter()

        Private secondCounter As PerformanceCounter = New PerformanceCounter()

        Public Sub New()
            InitializeComponent()
            FillCounterInformation()
            btnStart.Select()
        End Sub

        Private Sub FillCounterInformation()
            Dim processName As String = Process.GetCurrentProcess().ProcessName
            textEditCategory1.Text = "Process"
            textEditCounter1.Text = "Private Bytes"
            textEditInstance1.Text = processName
            textEditCategory2.Text = ".NET CLR Memory"
            textEditCounter2.Text = "# Bytes In All Heaps"
            textEditInstance2.Text = processName
        End Sub

        Private Function CreateCounters() As Boolean
            Dim p As Process = Process.GetCurrentProcess()
            Try
                firstCounter = New PerformanceCounter(textEditCategory1.Text, textEditCounter1.Text, textEditInstance1.Text)
                secondCounter = New PerformanceCounter(textEditCategory2.Text, textEditCounter2.Text, textEditInstance2.Text)
                firstCounter.NextValue()
                secondCounter.NextValue()
            Catch e As Exception
                MessageBox.Show(e.Message, "Counter Error")
                Return False
            End Try

            Return True
        End Function

        Private Sub btnStart_Click(ByVal sender As Object, ByVal e As EventArgs)
            If Not isRunning Then
                If Not CreateCounters() Then Return
                btnStart.Text = "Stop"
                dataProvider = New PerformanceDataProvider(firstCounter, secondCounter)
                AddHandler dataProvider.DataReady, AddressOf DataProvider_DataReady
                AddHandler dataProvider.DataOver, AddressOf DataProvider_DataOver
                isRunning = True
                dataProvider.Start(500)
            Else
                FullStop()
            End If
        End Sub

        Private Sub DataProvider_DataReady(ByVal sender As Object, ByVal e As EventArgs)
            sheetProcessor = New WorksheetProcessor(dataProvider.GetData())
            arcScaleComponent1.DataBindings.Add("Value", sheetProcessor.GetDataSource(), "Column 0")
            labelComponent1.DataBindings.Add("Text", sheetProcessor.GetDataSource(), "Column 0")
        End Sub

        Private Sub DataProvider_DataOver(ByVal sender As Object, ByVal e As EventArgs)
            FullStop()
        End Sub

        Private Sub FullStop()
            RemoveHandler dataProvider.DataReady, AddressOf DataProvider_DataReady
            RemoveHandler dataProvider.DataOver, AddressOf DataProvider_DataOver
            btnStart.Text = "Start"
            dataProvider.Stop()
            sheetProcessor.SaveData("mytest")
            isRunning = False
            arcScaleComponent1.DataBindings.Clear()
            labelComponent1.DataBindings.Clear()
        End Sub
    End Class
End Namespace
